app/jobs/autoscaling.py

```diff
@@ -260,227 +61,313 @@ def get_instance_tags():
+        return self.environment in set(s.strip() for s in allowed_envs.split(","))
+
+    @staticmethod
+    def find_attributes(asg_list, tag_list, instance_name):
+        def find_asg():
+            matching = []
+            for asg in asg_list:
+                if asg['AutoScalingGroupName'].startswith(instance_name):
+                    matching.append(asg)
+            assert len(matching) == 1
+            result = matching[0]
+            if DEBUG:
+                print json.dumps(result, indent=2)
+            return result
+        def find_instance_ids(asg):
+            return [item["InstanceId"] for item in asg["Instances"]]
+        def find_tags(instance_ids):
+            return [tag for tag in tag_list if tag["ResourceId"] in instance_ids]
+        asg = find_asg()
+        instance_ids = find_instance_ids(asg)
+        tags = find_tags(instance_ids)
+        return asg, instance_ids, tags
+
+    def num_desired_instances(self):
+        return int(self.asg.get("DesiredCapacity", self.asg.get("MinSize", 0)))
+
+    def tags_by_instance_id(self):
+        result = defaultdict(lambda: {})
+        for item in self.tags:
+            instance_id = item['ResourceId']
+            key = item['Key']
+            value = item['Value']
+            result[instance_id][key] = value
+        return result
+
+    def clamp_to_valid_range(self, desired_capacity):
+        min_size = int(self.asg.get("MinSize", desired_capacity))
+        max_size = int(self.asg.get("MaxSize", desired_capacity))
+        if desired_capacity < min_size:
+            print "Clamping {} up to {}".format(desired_capacity, min_size)
+            desired_capacity = min_size
+        if desired_capacity > max_size:
+            print "Clamping {} down to {}".format(desired_capacity, max_size)
+            desired_capacity = max_size
+        return desired_capacity
+
+    def set_desired_capacity(self, new_num_desired):
+        cmd = "aws autoscaling set-desired-capacity --auto-scaling-group-name {asg_name} --desired-capacity {new_num_desired}"
+        cmd = cmd.format(asg_name=self.asg['AutoScalingGroupName'], new_num_desired=new_num_desired)
+        if self.can_scale:
+            if DEBUG:
+                print cmd
+            aws_command(cmd)
+
+    def classify_instances(self):
+        healthy_instances = []
+        draining_instances = []
+        terminating_instances = []
+        corrupt_instances = []
+        tag_dict = self.tags_by_instance_id()
+        for inst in self.asg["Instances"]:
+            instance_id = inst["InstanceId"]
+            instance_tags = tag_dict.get(instance_id, {})
+            if DEBUG:
+                print "{instance_id}: {instance_tags}".format(instance_id=instance_id, instance_tags=instance_tags)
+            if not inst["ProtectedFromScaleIn"] and not inst["LifecycleState"] == "Pending":
+                terminating_instances.append(instance_id)
+            elif self.draining_tag in instance_tags:
```

Comment:
I suggest avoiding "elif" for something like this, because the conditions can pile up in difficult-to-comprehend ways.    Right now this entire statement is equivalent to the following;  are all those conjuncts correct?

```
if not inst["ProtectedFromScaleIn"] and not inst["LifecycleState"] == "Pending":
    terminating_instances.append(instance_id)

if (inst["ProtectedFromScaleIn"] or inst["LifecycleState"] == "Pending") and self.draining_tag in instance_tags:
    draining_instances.append(instance_id)

if (inst["ProtectedFromScaleIn"] or inst["LifecycleState"] == "Pending") and not self.draining_tag in instance_tags and inst["HealthStatus"] == "Healthy" and inst["LifecycleState"] != "Terminating":
    healthy_instances.append(instance_id)

if (inst["ProtectedFromScaleIn"] or inst["LifecycleState"] == "Pending") and not self.draining_tag in instance_tags and (inst["HealthStatus"] != "Healthy" or inst["LifecycleState"] == "Terminating")
    corrupt_instances.append(instance_id)
```

By that last one, it's difficult to follow exactly what condition makes an instance corrupt.   It would be best if that were simplified, avoiding the use of `elif` so these conjunct pileups don't lead to unintended consequences.

After that, you would wish to enforce mutual exclusion on the states.   If an instance occurs in more than one state, you can dump an error message, as it means logic error in this classifier.

Reply:
I'll say that all things considered it's very nice to have any such complexity encapsulate here in this function, so the rest of the program can deal in 4 simple mutually exclusive states.

Reply:
ok. I refactored to have the conditions for each state fully written out. I removed the "corrupt" state and replaced it with a check that each instance is in exactly 1 of the valid states.